### PR TITLE
GSoC 2019 Blog

### PR DESCRIPTION
**Pull request by 12Chao**
Hi guys, this is a blog of Cantera debugging tool PyCanChemYouCan2(ChemCheck). 
Pleas feel free to leave your comments to the project.

### CODE REVIEW

File: posts/GSoC 2019.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+## What is PyCanChemAndYouCan2(ChemCheck)
+
+PyCanChemAndYouCan2 is a web-based debugging tool developed for Cantera to visualize and diagnose syntax errors and chemical errors in input files. Users can upload  input files and visualize their errors on the website, and they can also download it after editing the input files.
+
+## Why Would I Need It
+
+Cantera uses input files containing mechanism of reactions because mechanism information is the fundamental to do thermo-kinetic calculation. The input files are required to be in .cti format, and users are able to gain them from various sources (e.g.: writing .cti file by hands, adopted from various sources and repositories on the web, or via a hybrid of these two approaches etc.).   However, .cti files always include errors since they are from different sources, which could cause  inaccurate results. The errors include syntax errors and chemical errors, and it is not likely to be found manually especially in large mechanism files. In this case, a debugging tool helping users to find and fix errors is very useful.
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: Please put an extra line after this paragraph before the header.

File: posts/GSoC 2019.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+## What is PyCanChemAndYouCan2(ChemCheck)
+
+PyCanChemAndYouCan2 is a web-based debugging tool developed for Cantera to visualize and diagnose syntax errors and chemical errors in input files. Users can upload  input files and visualize their errors on the website, and they can also download it after editing the input files.
+
+## Why Would I Need It
+
+Cantera uses input files containing mechanism of reactions because mechanism information is the fundamental to do thermo-kinetic calculation. The input files are required to be in .cti format, and users are able to gain them from various sources (e.g.: writing .cti file by hands, adopted from various sources and repositories on the web, or via a hybrid of these two approaches etc.).   However, .cti files always include errors since they are from different sources, which could cause  inaccurate results. The errors include syntax errors and chemical errors, and it is not likely to be found manually especially in large mechanism files. In this case, a debugging tool helping users to find and fix errors is very useful. 
+## What Errors Can Be Visualized or Fixed:
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: Please put an extra line after this header before the text.

File: posts/GSoC 2019.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+Syntax errors
+
+-   duplicate lines
+-   missing digits from end of line
+-   missing 'E'
+-   whatever else we find
+
+Chemistry errors
+
+-   discontinuous polynomicals
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: There should be one space after the `-` in any list.

File: posts/GSoC 2019.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+-   exceeding collision limit
+-   inconsistent thermo for explicit reverse rates
+-   negative rate coefficient (as a  [result](https://github.com/comocheng/wiki/issues/375#)  of sum of arrhenius or PLOG expressions, eg.  [here](https://github.com/Cantera/cantera-website/issues/77))
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: Arrhenius is someone's name, so should be capitalized

File: posts/GSoC 2019.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+-   excess stiffness
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: This one and the next two aren't really chemistry errors, so much as general problems. I would classify them into a new category.

Comment (bryanwweber) [reply posts/GSoC 2019.md]: You also need to run a simulation to catch them, and even then, the simulation would have to be run under some particular conditions. I'm not sure this is something the website can do very easily. However, making it easier to understand the errors in Cantera is very valuable.

File: posts/GSoC 2019.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+-   making  [error](https://github.com/comocheng/wiki/issues/375#)  messages from CVODES easier to understand
+-   making errors from CVODES easier to catch (raise a python Error)
+-   dead-end pathways
+## How Can I Achieve It
+
+The debugging tool PyCanChemAndYouCan2 is a web-based application developed in Django 2.2.  Users can upload original files and download revised files from the website, and the application contains  ACE code editor to allow users edit their files on the website. For Debugging script, Cantera has `ck2cti.py` file to convert different file format, so the debugging part will be test different defective mechanism files and update `ck2cti.py` file. The ideal condition is that the application can help users find the errors and sugget the fixes or fix errors automatically. However, it might be a long-term goal, so we make some minimum goal for this GSoC project.
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: What do you mean "update ck2cti.py file"? Are you planning to make changes to ck2cti? Or use ck2cti to identify conversion errors from Chemkin to CTI format. This paragraph isn't that clear. What is the long-term goal? What are the short-term goals? What will you have completed by the end of the GSoC program?

File: posts/GSoC 2019.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+Here are the basic ideas about the project:
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: You need one blank line before this line to make it a new paragraph.

File: posts/GSoC 2019.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ -   a script that takes in an input file and returns an input file with added comments, which describe detected errors.
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: You have an extra space in front of the `-` here.

File: posts/GSoC 2019.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+-   a web site that makes it easy to interact with that script (which runs on the web server), and renders the comments in an attractive way
+
+The idea is that the script can be used, developed, tested, as a stand-alone thing without the website.
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: This paragraph seems like it is copy-pasted from another comment. Can you make it into your own words?

File: posts/GSoC 2019.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+ -   a script that takes in an input file and returns an input file with added comments, which describe detected errors.
+-   a web site that makes it easy to interact with that script (which runs on the web server), and renders the comments in an attractive way
+
+The idea is that the script can be used, developed, tested, as a stand-alone thing without the website.  
+The website just makes things look prettier. eg. a custom "syntax highlighting" mode that detects comments describing certain problems and renders them in red, or whatever.
+
+Here is the workflow of the application:
+
+```
+A[upload files] --> B[list files] --> C{choose one and execute by ck2cte.py}  --> D{show input file with added comments on ACE editor}
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: Why do some of these have square brackets `[]` and some have curly brackets `{}`? What are the letters `A`, `B`, etc. for? You have a small typo, `ck2cte.py`

File: posts/GSoC 2019.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+--> E[edit file] --> F[download edited file] 
+```
+## How Much Have I Got
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: I would recommend waiting to describe your progress so far for the next update in 2 weeks. Then you can have a lot more to talk about.

File: posts/GSoC 2019.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+I have got the upload and download functions and merged ACE code editor into the web-application, the current challenge is how to display the files on the ACE code editor after the file has been executed by `ck2cti.py`.
+I am looking for a way to execute external python script in a Django-developed website, or how can I execute the python file using Javascript language. I am refering [django-ace]([https://github.com/django-ace/django-ace](https://github.com/django-ace/django-ace)), [RMGwebsite]([https://github.com/ReactionMechanismGenerator/RMG-website](https://github.com/ReactionMechanismGenerator/RMG-website)).
+
+My project code was posted on https://github.com/comocheng/ChemCheck, and I will update it every Sunday.
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: This is good to list the repository where the code will be located! However, I would very strongly suggest that you update the code far more often than only every Sunday, probably should be updated every day that you're working, and probably multiple times per day. I would really suggest that we define several features/checkpoints for you to implement, and that each of those should be completed in a pull request to the `ChemCheck` repository.

File: posts/GSoC 2019.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+Please leave your comments here if you have any suggestions for the project.
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]: We don't have comments enabled on the website :-) You could suggest that people post on Google Groups, I will (or you can) create a topic once this blog post goes live to announce the project.

File: posts/GSoC 2019.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
+---
+title: GSoC 2019 PyCanChemAndYouCan2
+date: 2019-06-07 12:33:00 UTC-16:33
+slug: GSoC 2019
+tags: GSoC 2019
+description:  2019 Google Summer of Code Cnatera Project Blog
```
[the remark this answers is not in the document]
Comment (bryanwweber) [reply posts/GSoC 2019.md]:
```suggestion
description: 2019 Google Summer of Code Cnatera Project Blog 
```

File: posts/GSoC 2019.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
+---
+title: GSoC 2019 PyCanChemAndYouCan2
+date: 2019-06-07 12:33:00 UTC-16:33
+slug: GSoC 2019
+tags: GSoC 2019
+description:  2019 Google Summer of Code Cnatera Project Blog 
+type: text
+author: Chao Xu
+
```
Comment (bryanwweber) [inline posts/GSoC 2019.md]:
```suggestion
```

File: posts/GSoC 2019.md (hunk 12 above)
Comment (bryanwweber) [reply posts/GSoC 2019.md]: But I would suggest that you do not mention the update frequency in the post.

File: posts/GSoC 2019.md (hunk 15 above)
Comment (12Chao) [reply posts/GSoC 2019.md]: Hi Bryan, thanks for your comments. I just revised my blog, especially the  "How can I achieve" part. I will meet with Rechard and make checkpoints next week.

**Comment by rwest** [comment]
Looks good to me. Thanks for your suggestions Bryan, and for implementing them Chao!

**Comment by bryanwweber** [comment]
Hi @12Chao Good work! I made some small grammar and spelling fixes. Please read it and make sure you agree with everything, once I get :+1: from you I will merge and the post will go live!

PS In the future, please add your posts on a new local branch for your work, you shouldn't make commits to `master`. It makes it harder for me to suggest changes when you commit directly to your `master`, but if you check out a new branch first, its very easy. Thanks!

**Comment by 12Chao** [comment]
Hi @bryanwweber , thanks for the changes, I reviewed and agree with the fixes.
Sorry for the inconvenience. I will add posts on another branch in the future.

**Pull request merged**